src/write_gcm.c

```diff
@@ -2275,59 +2274,155 @@ static monitored_resource_t *monitored_resource_create_from_fields(
+    wg_yajl_callback_size_t length) {
+  wg_mr_parse_context_t *ctx = (wg_mr_parse_context_t*) arg;
+
+  if(ctx->processing_labels) {
+    ctx->last_key = strndup((const char*) data, length);
+  } else if(strncmp((const char*) data, "labels", length) == 0) {
+    ctx->processing_labels = 1;
+  } else if(strncmp((const char*) data, "type", length) != 0) {
+    ERROR("wg_handle_map_key: Unexpected key in JSON.");
+    return 0;
+  }
+
+  return 1;
+}
+
+static int wg_mr_handle_end_map(void * arg) {
+  wg_mr_parse_context_t *ctx = (wg_mr_parse_context_t*) arg;
+
+  if(ctx->processing_labels == 1) {
+    ctx->processing_labels = 0;
+  }
+
+  return 1;
+}
+
 static monitored_resource_t *parse_monitored_resource(char *metadata,
     int metadata_size, const char *project_id) {
-  char errbuf[1024];
+  yajl_handle handle;
   monitored_resource_t *response = NULL;
-  yajl_val node = yajl_tree_parse(metadata, errbuf, sizeof(errbuf));
-  if (node == NULL) {
-    if (strlen(errbuf)) {
-      DEBUG("write_gcm: parse_monitored_resource %s", errbuf);
-    } else {
-      DEBUG("write_gcm: parse_monitored_resource: Invalid JSON response.");
-    }
-    goto finish;
-  }
+
   {
-    const char * type_path[] = {"type", (const char *) 0};
-    const char * labels_path[] = {"labels", (const char *) 0};
-    char *type;
-    yajl_val type_v = yajl_tree_get(node, type_path, yajl_t_string);
-    yajl_val labels_v = yajl_tree_get(node, labels_path, yajl_t_object);
-    if (type_v && YAJL_IS_STRING(type_v)) {
-      type = YAJL_GET_STRING(type_v);
-    } else {
-      ERROR("write_gcm: wg_parse_monitored_resource: %s not correctly defined.",
-            type_path[0]);
+    yajl_callbacks callbacks = {
+      .yajl_null = NULL,
+      .yajl_boolean = NULL,
+      .yajl_integer = NULL,
+      .yajl_double = NULL,
+      .yajl_number = NULL,
+      .yajl_string = &wg_mr_handle_string,
+      .yajl_start_map = NULL,
+      .yajl_map_key = &wg_mr_handle_map_key,
+      .yajl_end_map = &wg_mr_handle_end_map,
+      .yajl_start_array = NULL,
+      .yajl_end_array = NULL
+    };
+
+    wg_mr_parse_context_t parse_context = {
+      .processing_labels = 0,
+      .total_labels = 0
+    };
+
+    #if defined(YAJL_MAJOR) && YAJL_MAJOR >= 2
+      handle = yajl_alloc(&callbacks, NULL, &parse_context);
+    #else
+      yajl_parser_config config = { 0, 1 };
+      handle = yajl_alloc(&callbacks, &config, NULL, &parse_context);
+    #endif
+
+    if(yajl_parse(handle, (const unsigned char*) metadata, strlen(metadata)) != yajl_status_ok) {
+      ERROR("write_gcm: parse_monitored_resource: Invalid JSON response.");
       goto finish;
     }
-    size_t num_values;
-    if (labels_v && YAJL_IS_OBJECT(labels_v)) {
-      num_values = labels_v->u.object.len;
-    } else {
-      ERROR("write_gcm: wg_parse_monitored_resource: No such node %s.\n",
-            labels_path[0]);
+
+    #if defined(YAJL_MAJOR) && YAJL_MAJOR >= 2
+      int parse_result = yajl_complete_parse(handle);
```

Comment:
Could you add a comment specifying the difference between `yajl_parse` and `yajl_parse_complete` and why they are both needed?